### PR TITLE
Run checks in parallel

### PR DESCRIPTION
**Pull request by amitaekbote**

### CODE REVIEW

File: runner/runner.go [hunk 1]
```diff
@@ -231,60 +264,77 @@ func (r *Runner) run(ctx context.Context, checkMap map[string]*Check, list bool,
...
+	responseChannel := make(chan *Response, len(currentCheckList))
+	wg := &sync.WaitGroup{}
+	
 	// main loop to get the checks info.
 	for _, name := range currentCheckList {
-		resp := &Response{
-			name: name,
-		}
+		wg.Add(1)
+		xname := name
```
Comment (gpaul) [inline runner/runner.go]: If you do this you don't need to bind the variable as an argument to the closure on the next line - and conversely, if you pass `name` to the closure below on line 335, you don't need to do this local showing trick

File: runner/runner.go [hunk 2]
```diff
@@ -231,60 +264,77 @@ func (r *Runner) run(ctx context.Context, checkMap map[string]*Check, list bool,
...
+		go func(xname string, responseChannel chan *Response) {
+			defer wg.Done()
+			currentCheck, ok := checkMap[xname]
+			if !ok {
+				combinedResponse.errs["check not found"] = &Response{
```
Comment (gpaul) [inline runner/runner.go]: Pretty sure this is a race condition. You are assigning to `errs` from multiple goroutines. The coolest way to fix this IMO is to add a method on combinedResponse called `SetErr()` or similar and have that method take a mutex lock when it enters and defers the unlock.

File: runner/runner.go [hunk 3]
```diff
@@ -231,60 +264,77 @@ func (r *Runner) run(ctx context.Context, checkMap map[string]*Check, list bool,
...
+			if err != nil {
+				combinedResponse.errs[err.Error()] = resp
+			}
 
-		combinedResponse.status = max(combinedResponse.status, code)
+			combinedResponse.status = max(combinedResponse.status, code)
+			
+		}(xname, responseChannel)
 	}
+	wg.Wait()
```
Comment (gpaul) [inline runner/runner.go]: What do you think about the following alternative?

```
type responseError struct {
  err error
  checkName string
}

errs := make(chan responseError)
responses := make(chan error)

for _, name := range currentCheckList {
  go func(name string) {
    defer wg.Done()
    currentCheck, ok := checkMap[name]
    if !ok {
      errs <- responseError{errors.New("check not found"), name}
      return
    }
    ...
    // find runner for the given role only.
    if !currentCheck.verifyRole(r.role) {
      // send an empty response for this check, as it is not relevant for this role.
      responses <- nil
      return
    }
    ...
    resp := &Response{ ... }
    ...
    responses <- resp
    return
  }(name)
}

// Expect one response or error for every check
for range currentCheckList {
  select {
  case err := <-errs:
    combinedResponse.errs[err.Error()] = err.name
  case resp := <-responses:
    if resp == nil {
      // skip filtered checks that are not appropriate for this role.
      continue
    }
    combinedResponse.checks[resp.name] = resp
  }
}

combinedResponse.status = ...
return combinedResponse, nil
```

Comment (klueska) [reply runner/runner.go]: This is how I would do it

File: runner/runner.go [hunk 4]
```diff
@@ -231,59 +240,92 @@ func (r *Runner) run(ctx context.Context, checkMap map[string]*Check, list bool,
...
+				return
+			}
 
-		var (
-			combinedOutput []byte
-			code           int
-			err            error
-			checkDuration  string
-		)
+			if _, ok := combinedResponse.checks[name]; ok {
```
Comment (gpaul) [inline runner/runner.go]: Maybe we should move this check to the for loop by deduplicating `currentCheckList` before we start looping?

File: runner/runner.go [hunk 5]
```diff
@@ -55,6 +55,14 @@ type responseList struct {
 	Timeout     string   `json:"timeout"`
 }
 
+type responseError struct {
+	err           string
+	checkName     string
+	checkNotFound bool
+	response      *Response
+
```
Comment (mnaboka) [inline runner/runner.go]: redundant space

**Comment by gpaul** [comment]
@amitaekbote is this ready for another pass?

**Comment by amitaekbote** [comment]
yes it is

File: runner/runner.go [hunk 6]
```diff
@@ -231,59 +240,85 @@ func (r *Runner) run(ctx context.Context, checkMap map[string]*Check, list bool,
 		}
 	}
 
+	errs := make(chan responseError, len(currentCheckList))
+	responses := make(chan *Response, len(currentCheckList))
```
Comment (gpaul) [inline runner/runner.go]: As `responseError` contains a `*Response`, could we get rid of the `responses` channel and rename `errs` to `results` or something similar? Then we have only one channel. (I believe you suggested this a while ago.)

File: runner/runner.go (hunk 4 above)
Comment (gpaul) [reply runner/runner.go]: Checking `combinedResponse.checks` here is a data race: the on lines 303+ modify this map. I think we should just range through `currentCheckList` on line 246 (before entering the for loop) and remove duplicates.

File: runner/runner.go [hunk 7]
```diff
@@ -231,59 +240,85 @@ func (r *Runner) run(ctx context.Context, checkMap map[string]*Check, list bool,
...
+			// collect errors
+			if err != nil {
+				errs <- responseError{err.Error(), name, false, resp}
+			}
+		}(name)
+	}
+
+	for range currentCheckList {
+		select {
+		case err := <- errs:
```
Comment (gpaul) [inline runner/runner.go]: if we rename `errs` to `results` and remove `responses` in favour of using `resultError.response` then we can merge the two `case` statements here.

**Comment by branden** [comment]
I ported this PR to the new check runner component: https://github.com/dcos/dcos-check-runner/pull/3. @amitaekbote You can go ahead and close this out.

**Pull request closed without merging**